docs/designs/013-audit-logging.md

```diff
@@ -0,0 +1,479 @@
+    participant BG as Background Goroutine
+    participant Mongo as MongoDB
+    participant K8s as K8s Events API
+
+    Main->>Logger: Log(auditLog)
+    Logger->>Chan: Send (non-blocking)
+    Main->>Main: Continue execution
+    
+    Note over Main: Main operation unblocked
+    
+    BG->>Chan: Receive audit log
+    
+    par MongoDB Insert
+        BG->>Mongo: InsertAuditLog()
+        Mongo-->>BG: Success/Error
+    and K8s Event Create
+        BG->>K8s: CreateEvent()
+        K8s-->>BG: Success/Error
+    end
+    
+    Note over BG: Log error if failed,<br/>don't propagate to main
+```
+
+### Key Implementation Details
+
+#### 1. Non-Blocking Log Method
+
+```go
+func (l *Logger) Log(auditLog *AuditLog) {
+    if !l.config.Enabled {
+        return
+    }
+    
+    // Try to send to channel, but don't block if buffer is full
+    select {
+    case l.auditChan <- auditLog:
+        // Successfully queued
+    default:
+        // Buffer full - log warning and drop the audit log
+        slog.Warn("Audit log buffer full, dropping audit log",
+            "component", l.config.Component,
+            "operation", auditLog.Operation,
+            "resource", auditLog.ResourceName)
+    }
+}
+```
+
+#### 2. Graceful Error Handling
+
+```go
+func (l *Logger) processAuditLog(auditLog *AuditLog) {
+    // Log to MongoDB if enabled
+    if l.config.MongoDBEnabled && l.mongoClient != nil {
+        if err := l.mongoClient.InsertAuditLog(l.ctx, auditLog); err != nil {
+            slog.Error("Failed to insert audit log to MongoDB",
+                "error", err,
+                "component", l.config.Component)
+            // Continue - don't fail the main operation
+        }
+    }
+    
+    // Create Kubernetes event if enabled
+    if l.config.K8sEventsEnabled && l.k8sClient != nil {
+        if err := l.createK8sEvent(auditLog); err != nil {
+            slog.Error("Failed to create Kubernetes event",
+                "error", err)
+            // Continue - don't fail the main operation
+        }
+    }
+}
+```
+
+#### 3. Integration Pattern in Components
+
+```go
+// In component initialization
+func NewK8sClient(ctx context.Context, ...) (*Client, error) {
+    // Load config from environment variables
+    auditLogConfig := auditlog.ConfigFromEnv(auditlog.ComponentFaultQuarantine)
+    
+    auditLogger, err := auditlog.NewLogger(ctx, auditLogConfig, kubeClient)
+    if err != nil {
+        slog.Error("Failed to create audit logger", "error", err)
+        // Don't fail - audit is best-effort
+    }
+    
+    // Start async processor
+    if auditLogger != nil {
+        auditLogger.Start()
+    }
+    
+    return &Client{auditLogger: auditLogger}, nil
+}
+
+// In operation method
+func (c *Client) QuarantineNode(nodeName string) error {
+    // Perform the actual operation
+    err := c.k8sClient.CoreV1().Nodes().Patch(...)
+    if err != nil {
+        return err
+    }
+    
+    // Log audit (fire and forget)
```

Comment:
This is very invasive in the sense that we have to add this change everywhere we have client-go code. If we add new calls, but we forge to add this then audit logs will be missed. Can we instead think of a middleware or some kind of custom http client that will automatically audit log all the changes being made?